crates/bevy_ecs/src/error/handler.rs

```diff
@@ -77,51 +75,75 @@ impl ErrorContext {
-///
-/// This should be set inside of your `main` function, before initializing the Bevy app.
-/// The value of this error handler can be accessed using the [`default_error_handler`] function,
-/// which calls [`OnceLock::get_or_init`] to get the value.
-///
-/// **Note:** this is only available when the `configurable_error_handler` feature of `bevy_ecs` (or `bevy`) is enabled!
-///
-/// # Example
-///
-/// ```
-/// # use bevy_ecs::error::{GLOBAL_ERROR_HANDLER, warn};
-/// GLOBAL_ERROR_HANDLER.set(warn).expect("The error handler can only be set once, globally.");
-/// // initialize Bevy App here
-/// ```
-///
-/// To use this error handler in your app for custom error handling logic:
-///
-/// ```rust
-/// use bevy_ecs::error::{default_error_handler, GLOBAL_ERROR_HANDLER, BevyError, ErrorContext, panic};
-///
-/// fn handle_errors(error: BevyError, ctx: ErrorContext) {
-///    let error_handler = default_error_handler();
-///    error_handler(error, ctx);        
-/// }
-/// ```
-///
-/// # Warning
-///
-/// As this can *never* be overwritten, library code should never set this value.
-#[cfg(feature = "configurable_error_handler")]
-pub static GLOBAL_ERROR_HANDLER: OnceLock<fn(BevyError, ErrorContext)> = OnceLock::new();
+mod global_error_handler {
+    use super::{BevyError, ErrorContext};
+    use bevy_platform_support::sync::atomic::{
+        AtomicPtr,
+        Ordering::{AcqRel, Acquire, Relaxed},
+    };
+
+    // SAFETY: Only accessible from within this module.
+    static HANDLER_ADDRESS: AtomicPtr<()> = AtomicPtr::new(core::ptr::null_mut());
+
+    /// Set the global error handler.
+    ///
+    /// If used, this should be called [before] any uses of [`default_error_handler`],
+    /// generally inside your `main` function before initializing the app.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_ecs::error::{set_global_default_error_handler, warn};
+    /// set_global_default_error_handler(warn);
+    /// // initialize Bevy App here
+    /// ```
+    ///
+    /// To use this error handler in your app for custom error handling logic:
+    ///
+    /// ```rust
+    /// use bevy_ecs::error::{default_error_handler, BevyError, ErrorContext, panic};
+    ///
+    /// fn handle_errors(error: BevyError, ctx: ErrorContext) {
+    ///    let error_handler = default_error_handler();
+    ///    error_handler(error, ctx);        
+    /// }
+    /// ```
+    ///
+    /// # Warning
+    ///
+    /// As this can *never* be overwritten, library code should never set this value.
+    ///
+    /// [before]: https://doc.rust-lang.org/nightly/core/sync/atomic/index.html#memory-model-for-atomic-accesses
+    /// [`default_error_handler`]: super::default_error_handler
+    pub fn set_global_default_error_handler(handler: fn(BevyError, ErrorContext)) {
+        if HANDLER_ADDRESS
+            .compare_exchange(core::ptr::null_mut(), handler as *mut (), AcqRel, Acquire)
```

Comment:
Might we wan this to point to `panic` by default instead of null, allow it to be set multiple times, and get rid of the branch in `default_error_handler` and the nullability in `get`? I'm fine either way.

Reply:
Just to double check, can you walk me through why we would need `Acquire` here?  Function pointers are effectively `'static` right, is that not strong enough?

My other thought is, if we only want to write to this once then we don't actually need atomics do we? There are never conflicting updates, the problem is entirely memory ordering. So we can use an unsafe cell and fences, can't we? That would alleviate the concern about the function pointer size on exotic platforms.